Purchase with signature

friday-james opened the pull request.

friday-james committed:
remove unused variable

friday-james committed:
check range after resetting the index

friday-james committed:
validate promo code by owner address instead of discount percentage

friday-james committed:
purchase with signature

friday-james committed:
check requireSignature and totalPurchasedAmount

friday-james committed:
add purchase via promo code

friday-james committed:
reduce code size

friday-james committed:
add test case for signed purchases

friday-james committed:
update deployment script

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -208,101 +207,132 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable {
+        // the message has to be signed by operator
+        require(hasRole(OPERATOR_ROLE, signer), "Invalid signature");
+
+        require(getTotalPurchasedAmount(msg.sender) + _amount <= _allocation, "Purchase exceeds allocation");
```
Allocation changed to maxPaymentSent
GetTotalPurchasedAmoutn changed to GetTotalPaymentSent
_amount changed to amount * price in this tier

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -208,101 +207,132 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable {
-            require(purchasedAmountPerTier[_tierId][msg.sender] + _amount <= _allocation, "Purchase exceeds allocation");
-        }
+        require(tiers[_tierId].requireSignature, "Use whitelisted purchase");
+        bytes32 messageHash = keccak256(abi.encodePacked(msg.sender, _tierId, _allocation));
```
Needs to check contract address / payment token decimal/address

friday-james committed:
check payment token received instead of node purchased

friday-james committed:
update compile script

friday-james committed:
check payment amount per contract

friday-james committed:
pass promo code instead of walletPromoCode

friday-james committed:
Merge branch 'audit-review' into signed-purchase

friday-james committed:
avoid unexpected reverts due to underflows

friday-james committed:
audit review:
- fix compilation error
- check chainId in messageHash
- check wallet code in signed purchase
- make sure promo code exists on updatePromoCode

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -1,26 +1,25 @@
 import "./IFFundable.sol";
 
 // Contract to manage tiered sales with promotional codes and whitelisting.
-contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable {
```
why do we remove ReentrancyGuard?
